### PR TITLE
Testing Spatial/Temporal indexes

### PR DESCRIPTION
**Pull request by sherfert**
This PR adds unit tests and ITs around spatial and temporal indexes.

It also addresses some bugs that were uncovered by the new tests.

### CODE REVIEW

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DurationUniqueSchemaIndexAccessorTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.neo4j.values.storable.ValueGroup;
+
+import static org.neo4j.index.internal.gbptree.RecoveryCleanupWorkCollector.IMMEDIATE;
+
+public class DurationUniqueSchemaIndexAccessorTest extends NativeSchemaIndexAccessorTest<DurationSchemaKey,NativeSchemaValue>
+{
+    @Override
+    NativeSchemaIndexAccessor<DurationSchemaKey,NativeSchemaValue> makeAccessorWithSamplingConfig( IndexSamplingConfig samplingConfig ) throws IOException
+    {
+        TemporalIndexFiles.FileLayout<DurationSchemaKey> fileLayout = new TemporalIndexFiles.FileLayout<>( getIndexFile(), layout, ValueGroup.LOCAL_DATE_TIME );
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DurationUniqueSchemaIndexAccessorTest.java]: Copy pasta: `ValueGroup.LOCAL_DATE_TIME`, should be `ValueGroup.DURATION`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DurationLayoutTestUtil.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import org.neo4j.test.rule.RandomRule;
+import org.neo4j.values.storable.DurationValue;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.Values;
+
+public class DurationLayoutTestUtil extends LayoutTestUtil<DurationSchemaKey, NativeSchemaValue>
+{
+    private static final DurationValue[] ALL_EXTREME_VALUES = new DurationValue[]
+    {
+            DurationValue.duration( -999999999L * 12, 0, 0, 0),
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DurationLayoutTestUtil.java]: If you want to test the extreme values for duration, you should multiply the values in these four lines with 2. This because people can write e.g. `duration.between(LocalDate.of( -999999999, 1, 1), LocalDate.of( 999999999, 12, 31)) ` .

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        PointValue pointValue;
+        Long compareValue;
+        do
+        {
+            x = random.nextDouble() * 360 - 180;
+            y = random.nextDouble() * 180 - 90;
+            pointValue = Values.pointValue( crs, x, y );
+            compareValue = curve.derivedValueFor( pointValue.coordinate() );
+        }
+        while ( !uniqueCompareValues.add( compareValue ) );
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java]: I'm not entirely sure how the uniqueness constraint for Points was implemented, but does this mean that we will disregard points which are unique but happens to have the same 1D-value as an existing point? For me that sounds like an important case to test.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialUniqueSchemaIndexPopulatorTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+{
+    private static final CoordinateReferenceSystem crs = CoordinateReferenceSystem.WGS84;
+    private static final SpaceFillingCurveSettings settings = new SpaceFillingCurveSettingsFactory( Config.defaults() ).settingsFor( crs );
+
+    private SpatialIndexFiles.SpatialFileLayout fileLayout;
+
+    @Override
+    NativeSchemaIndexPopulator<SpatialSchemaKey,NativeSchemaValue> createPopulator( IndexSamplingConfig samplingConfig )
+    {
+        fileLayout = new SpatialIndexFiles.SpatialFileLayout( CoordinateReferenceSystem.WGS84, settings, super.getIndexFile() );
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialUniqueSchemaIndexPopulatorTest.java]: Maybe use your final variable `crs` here instead of `CoordinateReferenceSystem.WGS84`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialNonUniqueSchemaIndexPopulatorTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+{
+    private static final CoordinateReferenceSystem crs = CoordinateReferenceSystem.WGS84;
+    private static final SpaceFillingCurveSettings settings = new SpaceFillingCurveSettingsFactory( Config.defaults() ).settingsFor( crs );
+
+    private SpatialIndexFiles.SpatialFileLayout fileLayout;
+
+    @Override
+    NativeSchemaIndexPopulator<SpatialSchemaKey,NativeSchemaValue> createPopulator( IndexSamplingConfig samplingConfig )
+    {
+        fileLayout = new SpatialIndexFiles.SpatialFileLayout( CoordinateReferenceSystem.WGS84, settings, super.getIndexFile() );
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialNonUniqueSchemaIndexPopulatorTest.java]: Maybe use your final variable `crs` here instead of `CoordinateReferenceSystem.WGS84`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java [hunk 6]
```diff
@@ -697,7 +697,7 @@ public void shouldNotSeeFilteredEntries() throws Exception
 
         // when
         NodeValueIterator iter = new NodeValueIterator();
-        IndexQuery.ExactPredicate filter = IndexQuery.exact( 0, valueOf( updates[1] ) );
+        IndexQuery.ExactPredicate filter = IndexQuery.exact( 0, updates[1].values()[0].asObjectCopy() );
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java]: Why not using `valueOf( updates[1] ).asObjectCopy()` ?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java [hunk 7]
```diff
@@ -730,7 +730,7 @@ private Value generateUniqueValue( IndexEntryUpdate<SchemaIndexDescriptor>[] upd
 
     private Object valueOf( IndexEntryUpdate<SchemaIndexDescriptor> update )
     {
-        return update.values()[0].asObject();
+        return update.values()[0];
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java]: The method can now have the return type `Value`, which is more specific than `Object`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulatorTest.java [hunk 8]
```diff
@@ -599,7 +598,8 @@ private void verifyUpdates( Iterator<IndexEntryUpdate<SchemaIndexDescriptor>> in
     private byte[] fileWithContent() throws IOException
     {
         int size = 1000;
-        try ( StoreChannel storeChannel = fs.create( indexFile ) )
+        fs.mkdirs( getIndexFile().getParentFile() );
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulatorTest.java:601]: Out of curiosity, why was this addition necessary?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/UniqueLayoutTestUtil.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Override
+    IndexEntryUpdate<SchemaIndexDescriptor>[] someUpdates()
+    {
+        return delegate.someUpdatesNoDuplicateValues();
+    }
+
+    @Override
+    IndexQuery rangeQuery( Object from, boolean fromInclusive, Object to, boolean toInclusive )
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/UniqueLayoutTestUtil.java]: This method (+ its super method in `LayoutTestUtil` and its sub-methods) could change the signature to `IndexQuery rangeQuery( Value from, boolean fromInclusive, Value to, boolean toInclusive )`. See also comment about `valueOf`.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/LayoutTestUtil.java [hunk 10]
```diff
@@ -54,7 +54,10 @@
 
     abstract IndexEntryUpdate<SchemaIndexDescriptor>[] someUpdates();
 
-    protected abstract double fractionDuplicates();
+    protected double fractionDuplicates()
+    {
+        return 0.1;
```
Comment (Lojjs) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/LayoutTestUtil.java:59]: `StringNonUniqueLayoutTestUtil` and `NumberNonUniqueLayoutTestUtil` override this method with exact the same implementation as here so those methods can be deleted.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexProvider.java [hunk 11]
```diff
@@ -140,8 +141,15 @@ public InternalIndexState getInitialState( long indexId, SchemaIndexDescriptor d
     {
         SpatialIndexFiles spatialIndexFiles = new SpatialIndexFiles( directoryStructure(), indexId, fs, settingsFactory );
 
+        final Iterable<SpatialIndexFiles.SpatialFileLayout> existing = spatialIndexFiles.existing();
+        if ( !existing.iterator().hasNext() )
+        {
+            monitor.failedToOpenIndex( indexId, descriptor, "Requesting re-population.",
+                    new NoSuchFileException( spatialIndexFiles.forCrs( CoordinateReferenceSystem.WGS84 ).indexFile.getAbsolutePath() ) );
+            return InternalIndexState.POPULATING;
```
Comment (Lojjs) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexProvider.java:149]: Shouldn't we get `InternalIndexState.FAILED` after the exception on the line above? Same question also for  `TemporalIndexProvider`

File: community/spatial-index/src/test/java/org/neo4j/gis/spatial/index/EnvelopeTest.java [hunk 12]
```diff
@@ -106,6 +107,49 @@ public void shouldHandleIntersectionsIn2D()
...
+        Envelope envelope = new Envelope( from, to ).withSideRatioNotTooSmall();
+        double[] expectedFrom = new double[] {0, 0, 0, 0};
+        double[] expectedTo = new double[] {1, 1, 1, 1};
+        assertArrayEquals( expectedFrom, envelope.min , 0);
+        assertArrayEquals( expectedTo, envelope.max, 0 );
+
+        // Expected to change
+        to = new double[] {100, 0.1, 12, 0.01};
+        Envelope envelope2 = new Envelope( from, to ).withSideRatioNotTooSmall();
+        double[] expectedTo2 = new double[] {100, 1, 12, 1};
```
Comment (Lojjs) [inline community/spatial-index/src/test/java/org/neo4j/gis/spatial/index/EnvelopeTest.java]: In both this and the test above all changed values are expected to become 1, due to the highest diff being 100. Maybe we should test with some other value as well.

**Review by Lojjs — CHANGES_REQUESTED** [review]
Good with more testing :) Some mixed comments though

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java (hunk 3 above)
Comment (sherfert) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java]: You're right. Unfortunately the test framework only exercises that part of the code, without the subsequent filtering that is needed for Spatial. That means if we allow Points that map to the same 1D-value here, the tests are failing because they do not expect to get more than one result for a query, but get more. The framework is not really designed for lossy indexes, but still provides _some_ nice coverage for Spatial.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java (hunk 6 above)
Comment (sherfert) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java]: Yeah, that's right.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulatorTest.java (hunk 8 above)
Comment (sherfert) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulatorTest.java:601]: The structure of the indexes somehow got deeper (with a folder for the provider (e.g. lucene-1.0) and a subfolder for the sub-provider (e.g. temporal-1.0). Creating the file of these deeper nested directories requires creating the directories first. This was not true for Number/string/lucene, since they don't have sub-providers.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/UniqueLayoutTestUtil.java (hunk 9 above)
Comment (sherfert) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/UniqueLayoutTestUtil.java]: That's nice!

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexProvider.java (hunk 11 above)
Comment (sherfert) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexProvider.java:149]: Not according to this test:

```
@Test
    public void shouldReportInitialStateAsPopulatingIfIndexDoesntExist()
    {
        // given
        provider = newProvider();

        // when
        InternalIndexState state = provider.getInitialState( indexId, descriptor() );

        // then
        assertEquals( InternalIndexState.POPULATING, state );
        logging.assertContainsLogCallContaining( "Failed to open index" );
    }
```

It seems to be consistent with the other indexes, but I agree it's weird.

**Review by Lojjs — APPROVED** [review]

File: community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java [hunk 13]
```diff
@@ -174,6 +175,20 @@ private int countIndexedNodes( GraphDatabaseAPI db, Label label, String... keys
...
+                long runningFor;
+                InternalIndexState state;
+                do
+                {
+                    state = ktx.schemaRead().indexGetState( index );
+                    runningFor = System.currentTimeMillis() - start;
+                }
+                while ( state != InternalIndexState.ONLINE && runningFor < timeout );
+
+
```
Comment (chrisvest) [inline community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java]: There is also `db.schema().awaitIndexesOnline(…)`.

Comment (sherfert) [reply community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java]: I think I don't have access to that from this IT, do I?

Comment (sherfert) [reply community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java]: Yeah, let me use that one, I haven't seen that the GraphDatabaseAPI was in scope

Comment (sherfert) [reply community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java]: I tried, but somehow that does not actually seem to wait for that index...
